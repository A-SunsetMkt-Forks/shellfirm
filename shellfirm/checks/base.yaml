- from: base
  test: ':\(\)\{ :|:& \};:'
  description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack."
  id: base:bash_fork_bomb
- from: base
  test: \s*crontab\s+-r
  description: "You are going to remove your entire table of cron tasks."
  id: base:delete_all_cron_tasks
- from: base
  test: \s*history(.*)[|](.*)(bash|sh)($|\s)
  description: "You are going to executes every command from the command log that you have already executed."
  id: base:execute_all_history_commands
- from: base
  test: reboot(\s|$)
  description: "You are going to reboot your machine."
  id: base:reboot_machine
- from: base
  test: shutdown(\s|$)
  description: "You are going to shutdown your machine."
  id: base:shutdown_machine
