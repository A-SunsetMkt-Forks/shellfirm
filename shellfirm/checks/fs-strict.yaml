- from: fs-strict
  test: chmod\s
  description: "Are you sure that you want to change permissions?"
  id: fs-strict:change_permissions
- from: fs-strict
  test: rm\s
  description: "Are you sure you want to continue with deletion?"
  id: fs-strict:any_deletion