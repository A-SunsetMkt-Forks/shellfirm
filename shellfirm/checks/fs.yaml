- from: fs
  test: 'rm\s{1,}(-r|-fr|-rf)\s*(\*|\.{2,}|/)'
  method: Regex
  description: "You are going to delete everything in the path."
  id: fs:recursively_delete
- from: fs
  test: mv\s{1,}([a-zA-Z0-9.!@#$%^&*()']*) /dev/null
  method: Regex
  description: "The files will be discarded and destroyed."
  id: fs:move_to_dev_null
- from: fs
  test: .*>(.*)
  method: Regex
  description: "The above command is used to flush the content of a file."
  filters:
    IsFileExists: "1"
  id: fs:flush_file_content
- from: fs
  test: chmod\s{1,}(-R|--recursive)\s{1}(\*|\.{2,}|/)
  method: Regex
  description: "Change permission to all root files can brake your some thinks like SSH keys."
  id: fs:recursively_chmod
- from: fs
  test: find\s.*-delete
  method: Regex
  description: "Did you -delete flag in the wrong order? find -delete going to delete all the file under your current path."
  id: fs:delete_find_files
