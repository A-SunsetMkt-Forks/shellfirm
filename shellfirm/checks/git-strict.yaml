- from: git-strict
  test: git\s{1,}add\s{1,}.
  description: "Do you want to add all changes files to the staging area?"
  id: git-strict:git_add_all
- from: git-strict
  test: git\s{1,}commit\s{1,}(--all|-(.*a))
  description: "Do you want to commit all changes file and commit them?"
  id: git-strict:git_commit_all