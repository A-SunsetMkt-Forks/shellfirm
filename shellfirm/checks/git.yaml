- from: git
  test: git reset
  method: Contains
  description: "This command going to reset all your local changes."
  id: git:git_reset