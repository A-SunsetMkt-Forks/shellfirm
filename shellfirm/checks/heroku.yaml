- from: heroku
  test: heroku ps:restart
  method: Regex
  enable: true
  description: "Restarting app dynos"
- from: heroku
  test: heroku ps:stop
  method: Regex
  enable: true
  description: "Stop app dynos"
- from: heroku
  test: heroku ps:kill
  method: Regex
  enable: true
  description: "Kill app dyno"
- from: heroku
  test: heroku ps:kill
  method: Regex
  enable: true
  description: "Kill app dyno"
- from: heroku
  test: heroku maintenance:on
  method: Regex
  enable: true
  description: "Put the app into maintenance mode?"
- from: heroku
  test: heroku members:remove
  method: Regex
  enable: true
  description: "Removes a user from a team"
- from: heroku
  test: heroku features:disable
  method: Regex
  enable: true
  description: "Disables an app feature"
- from: heroku
  test: heroku container:rm
  method: Regex
  enable: true
  description: "Remove the process type from your app"
- from: heroku
  test: heroku config:unset
  method: Regex
  enable: true
  description: "unset one or more config vars"
- from: heroku
  test: heroku clients:destroy
  method: Regex
  enable: true
  description: "Delete client by ID"
- from: heroku
  test: heroku clients:rotate
  method: Regex
  enable: true
  description: "Rotate OAuth client secret"
- from: heroku
  test: heroku clients:update
  method: Regex
  enable: true
  description: "Update OAuth client"
- from: heroku
  test: heroku apps:destroy
  method: Regex
  enable: true
  description: "Permanently destroy an app"
- from: heroku
  test: heroku apps:leave
  method: Regex
  enable: true
  description: "Remove yourself from a team app"
- from: heroku
  test: heroku apps:rename
  method: Regex
  enable: true
  description: "Rename an app"
- from: heroku
  test: heroku addons:destroy
  method: Regex
  enable: true
  description: "Permanently destroy an add-on resource"
- from: heroku
  test: heroku addons:detach
  method: Regex
  enable: true
  description: "Detach an existing add-on resource from an app"
- from: heroku
  test: heroku access:remove
  method: Regex
  enable: true
  description: "Remove users from a team app"
- from: heroku
  test: heroku access:update
  method: Regex
  enable: true
  description: "Update existing collaborators on an team app"
- from: heroku
  test: heroku repo:reset
  method: Regex
  enable: true
  description: "Reset heroku repo"