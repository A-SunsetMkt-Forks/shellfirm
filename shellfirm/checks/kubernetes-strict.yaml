- from: kubernetes-strict
  test: \s*(k|kubectl)\s*(.*)delete[[:space:]][^ns]
  method: Regex
  enable: true
  description: "This command will going to delete a given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*set
  method: Regex
  enable: true
  description: "This command will update the given given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*scale
  method: Regex
  enable: true
  description: "This command will set a new size for a given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*rollout
  method: Regex
  enable: true
  description: "This command will manage a rollout for a given resource."