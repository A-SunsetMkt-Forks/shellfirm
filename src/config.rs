//! Configuration management

use crate::checks::Check;
use crate::cli::{UPDATE_CONFIGURATION_ONLY_DIFF, UPDATE_CONFIGURATION_OVERRIDE};
use anyhow::anyhow;
use anyhow::Result as AnyResult;
use serde_derive::{Deserialize, Serialize};
use std::fs;
use std::io::{Read, Write};

pub const DEFAULT_CONFIG_FILE: &str = include_str!("config.yaml");

/// The method type go the check.
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum Method {
    /// If the command start with.
    StartWith,
    /// if the command contains.
    Contains,
    /// if the command match to the given regex.
    Regex,
}

/// The user challenge when user need to confirm the command.
#[derive(Debug, Deserialize, Serialize)]
pub enum Challenge {
    /// Math challenge.
    Math,
    /// Only enter will approve the command.
    Enter,
    /// only yes/no typing will approve the command.
    YesNo,
}

/// describe configuration folder
pub struct SettingsConfig {
    /// Configuration folder path.
    pub path: String,
    /// config file.
    pub config_file_path: String,
    /// If configuration path overridden by the user.
    pub default: bool,
}

/// Describe the configuration yaml
#[derive(Debug, Deserialize, Serialize)]
pub struct Config {
    /// Type of the challenge.
    pub challenge: Challenge,
    /// List of checks.
    pub checks: Vec<Check>,
}

// /// Describe single check
// #[derive(Debug, Deserialize, Clone)]
// pub struct Check {
//     pub is: String,
//     pub method: Method,
//     pub enable: bool,
//     pub description: String,
// }

impl SettingsConfig {
    /// Convert config yaml file to struct.
    pub fn load_config_from_file(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(&self.read_config_file()?)?)
    }

    /// Return default app config as a config struct.
    pub fn load_default_config(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(DEFAULT_CONFIG_FILE)?)
    }

    /// Manage configuration folder/ file.
    /// * Create config folder if not exists
    /// * Create config yaml file if not exists
    pub fn manage_config_file(&self) -> AnyResult<()> {
        self.create_config_folder()?;
        if fs::metadata(&self.config_file_path).is_err() {
            self.create_default_config_file()?;
        }
        Ok(())
    }

    pub fn update_config_content(&self, behavior: &str) -> AnyResult<()> {
        match behavior {
            UPDATE_CONFIGURATION_ONLY_DIFF => self.add_diff_configuration_file(),
            UPDATE_CONFIGURATION_OVERRIDE => self.override_configuration_file(),
            _ => return Err(anyhow!("unsupported behavior")),
        }
    }

    /// Create config folder if not exists.
    fn create_config_folder(&self) -> AnyResult<()> {
        if let Err(err) = fs::create_dir(&self.path) {
            if err.kind() != std::io::ErrorKind::AlreadyExists {
                return Err(anyhow!("could not create folder: {}", err.to_string()));
            }
        }
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_default_config_file(&self) -> AnyResult<()> {
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(DEFAULT_CONFIG_FILE.as_bytes())?;
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_config_file_from_struct(&self, config: &Config) -> AnyResult<()> {
        let content = serde_yaml::to_string(config)?;
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }

    /// Convert config file to config struct.
    fn read_config_file(&self) -> AnyResult<String> {
        let mut file = std::fs::File::open(&self.config_file_path)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        Ok(content)
    }

    /// override existing config yaml file with the application configuration
    fn override_configuration_file(&self) -> AnyResult<()> {
        self.create_default_config_file()
    }

    /// adding only new check into configuration file.
    fn add_diff_configuration_file(&self) -> AnyResult<()> {
        match self.load_config_from_file() {
            Ok(mut conf) => {

                let default_app_config =  self.load_default_config()?;

                for default_check in default_app_config.checks.clone() {
                    let mut found = false;
                    for user_check in &conf.checks{
                        if user_check.is == default_check.is {found = true; break;}
                    }
                    if !found {
                        conf.checks.push(default_check.clone())
                    }
                }
                self.create_config_file_from_struct(&conf)?;
                Ok(())
            },
            Err(_e) => return Err(anyhow!("could not parse current config file. please try to fix the yaml file or override the current configuration by use the flag `--behavior override`"))
        }
    }
}

/// Get config config application details.
///
/// # Arguments
///
/// * `path` - Config folder path. if is empty default path will be returned.
pub fn get_config_folder(path: &str) -> AnyResult<SettingsConfig> {
    let package_name = env!("CARGO_PKG_NAME");

    let mut config_folder = path.into();
    let mut is_default = false;
    if path.is_empty() {
        match home::home_dir() {
            Some(path) => {
                is_default = true;
                config_folder = format!("{}/.{}", path.display(), package_name);
            }
            None => return Err(anyhow!("could not get directory path")),
        }
    }
    Ok(SettingsConfig {
        path: config_folder.clone(),
        default: is_default,
        config_file_path: format!("{}/config.yaml", config_folder),
    })
}

#[cfg(test)]
mod config {
    use super::*;

    fn get_current_project_path() -> String {
        std::env::current_dir().unwrap().to_str().unwrap().into()
    }
    #[test]
    fn can_get_config_default_folder() {
        let conf = get_config_folder("").unwrap();
        assert_ne!(conf.path, "");
        assert!(conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_get_config_custom_folder() {
        let folder_path = "/custom/folder/path";
        let conf = get_config_folder(folder_path).unwrap();
        assert_eq!(conf.path, folder_path);
        assert!(!conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/src/config.yaml", get_current_project_path()),
        };

        assert!(settings_config.load_config_from_file().is_ok())
    }

    #[test]
    fn cant_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: String::from(""),
            default: false,
            config_file_path: String::from(""),
        };

        assert!(settings_config.load_config_from_file().is_err())
    }

    #[test]
    fn can_write_config_file() {
        let tmp_folder = format!("{}/tmp", get_current_project_path());
        let config_file_path = format!("{}/config.yaml", &tmp_folder);
        if fs::metadata(&config_file_path).is_ok() {
            fs::remove_file(&config_file_path).unwrap();
        }

        let settings_config = SettingsConfig {
            path: format!("{}", tmp_folder),
            default: false,
            config_file_path: config_file_path,
        };

        assert!(settings_config.manage_config_file().is_ok());
        assert!(settings_config.read_config_file().is_ok());
    }

    #[test]
    fn can_load_default_config() {
        let conf = get_config_folder("").unwrap();
        assert!(conf.load_default_config().is_ok())
    }

    #[test]
    fn can_override_existing_config() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/override.yaml", get_current_project_path()),
        };

        // check if we file created successfully
        assert!(fs::File::create(&settings_config.config_file_path)
            .unwrap()
            .write_all("".as_bytes())
            .is_ok());

        // then read the file and make sure that the content is empty
        let file_content = settings_config
            .read_config_file()
            .unwrap_or(format!("error"));

        assert_eq!(file_content, "");

        // create the default configuration
        assert!(settings_config.override_configuration_file().is_ok());
        // make sure that the file is not empty
        assert!(!settings_config.read_config_file().unwrap().is_empty());
    }

    #[test]
    fn can_add_diff_configuration_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/add-diff.yaml", get_current_project_path()),
        };

        let orig_config = settings_config.load_default_config().unwrap();
        let mut config = settings_config.load_default_config().unwrap();

        // creates configuration file with only 1 check (we want to check that we not change existing check and just append new ones)
        config.checks = vec![Check {
            is: String::from("is value"),
            method: Method::Contains,
            enable: true,
            description: String::from("description"),
        }];

        // create configuration file with 1 check
        assert!(settings_config
            .create_config_file_from_struct(&config)
            .is_ok());

        // make sure that the configuration created with 1 check
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len()
        );

        // create the config diff command
        assert!(settings_config.add_diff_configuration_file().is_ok());

        // make sure that the count of check that change equal to existing config + the default application config
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len() + orig_config.checks.len()
        );
    }
}
