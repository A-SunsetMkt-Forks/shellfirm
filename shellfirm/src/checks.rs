//! Manage command checks

///
use crate::config::{Challenge, Method};
use crate::prompt;
use anyhow::Result;
use console::style;
use log::debug;
use rayon::prelude::*;
use regex::Regex;
use serde_derive::{Deserialize, Serialize};
use std::collections::HashMap;

// list of custom filter
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Hash, Clone)]
pub enum FilterType {
    IsFileExists,
}

/// Describe single check
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Check {
    /// test ia a value that we check the command.
    pub test: String,
    /// The type of the check.
    pub method: Method,
    /// boolean for ignore check
    pub enable: bool,
    /// description of what is risky in this command
    pub description: String,
    /// the group of the check see files in `checks` folder
    pub from: String,
    #[serde(default)]
    pub challenge: Challenge,
    #[serde(default)]
    pub filters: HashMap<FilterType, String>,
}

/// prompt a challenge to the user
///
/// # Errors
///
/// Will return `Err` when could not convert checks to yaml
pub fn challenge(challenge: &Challenge, checks: &[Check], dryrun: bool) -> Result<bool> {
    if dryrun {
        eprintln!("{}", serde_yaml::to_string(checks)?);
        return Ok(true);
    }
    eprintln!("{}", style("#######################").yellow().bold());
    eprintln!("{}", style("# RISKY COMMAND FOUND #").yellow().bold());
    eprintln!("{}", style("#######################").yellow().bold());

    let mut descriptions: Vec<String> = Vec::new();
    for check in checks {
        if !descriptions.contains(&check.description) {
            descriptions.push(check.description.to_string());
        }
    }
    for description in descriptions {
        eprintln!("* {}", description);
    }
    eprintln!();

    let show_challenge = challenge;

    Ok(match show_challenge {
        Challenge::Math | Challenge::Default => prompt::math_challenge(),
        Challenge::Enter => prompt::enter_challenge(),
        Challenge::Yes => prompt::yes_challenge(),
    })
}

/// Check if the given command matched to on of the checks
///
/// # Arguments
///
/// * `checks` - List of checks that we want to validate.
/// * `command` - Command check.
#[must_use]
pub fn run_check_on_command(checks: &[Check], command: &str) -> Vec<Check> {
    checks
        .par_iter()
        .filter(|&v| v.enable)
        .filter(|&v| is_match(v, command))
        .filter(|&v| check_custom_filter(v, command))
        .map(std::clone::Clone::clone)
        .collect()
}

/// filter custom checks
///
/// When true is returned it mean the filter should keep the check and not filter our the check.
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - Command.
fn check_custom_filter(check: &Check, command: &str) -> bool {
    // if check is not regex type or custom filters are not configure return true.
    if check.method != Method::Regex || check.filters.is_empty() {
        return true;
    }
    // Capture command groups from the current check
    let caps = Regex::new(&check.test).unwrap().captures(command).unwrap();

    // by default true is return. it mean the check not filter out (safe side security).
    let mut keep_check = true;
    for (filter_type, filter_params) in &check.filters {
        debug!(
            "filter information: command {} include filter: {:?} filter_params: {}",
            command, filter_type, filter_params
        );

        let keep_filter = match filter_type {
            FilterType::IsFileExists => filter_is_file_exists(
                caps.get(filter_params.parse().unwrap())
                    .map_or("", |m| m.as_str()),
            ),
        };

        if !keep_filter {
            keep_check = false;
            break;
        }
    }

    keep_check
}

/// Check if the given command match to one of the existing checks
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - command for the check
fn is_match(check: &Check, command: &str) -> bool {
    match check.method {
        Method::Contains => is_contains(&check.test, command),
        Method::StartWith => is_start_with(&check.test, command),
        Method::Regex => is_regex(&check.test, command),
    }
}

/// Is the command contains the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_contains(test: &str, command: &str) -> bool {
    command.contains(test)
}

/// is the command start with the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_start_with(test: &str, command: &str) -> bool {
    command.starts_with(test)
}

/// Is the command match to the given regex.
///
/// # Arguments
///
/// * `test_r` - check value
/// * `command` - command value
fn is_regex(test_r: &str, command: &str) -> bool {
    Regex::new(test_r).unwrap().is_match(command)
}

/// check if the path exists (file and folder).
///
/// # Arguments
///
/// * `file_path` - check path.
fn filter_is_file_exists(file_path: &str) -> bool {
    let mut file_path: String = file_path.trim().into();
    if file_path.starts_with('~') {
        match dirs::home_dir() {
            Some(path) => {
                file_path = file_path.replace('~', &path.display().to_string());
            }
            None => return true,
        };
    }
    debug!("check is file {} exists", file_path);
    return std::path::Path::new(file_path.trim()).exists();
}

#[cfg(test)]
mod test_checks {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn is_match_command() {
        let regex_check = Check {
            test: String::from("rm.+(-r|-f|-rf|-fr)*"),
            method: Method::Regex,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let contains_check = Check {
            test: String::from("test"),
            method: Method::Contains,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let startwith_check = Check {
            test: String::from("start"),
            method: Method::StartWith,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        assert_debug_snapshot!(&regex_check);
        assert_debug_snapshot!(&contains_check);
        assert_debug_snapshot!(&startwith_check);
    }
    #[test]
    fn can_check_is_contains() {
        assert_debug_snapshot!(is_contains("test", "test is valid"));
        assert_debug_snapshot!(is_contains("test is valid", "not-found"));
    }

    #[test]
    fn can_check_is_start_with() {
        assert_debug_snapshot!(is_start_with("test is", "test is valid"));
        assert_debug_snapshot!(is_start_with("test is valid", "is"));
    }

    #[test]
    fn can_check_is_regex_match() {
        assert_debug_snapshot!(is_regex("rm.+(-r|-f|-rf|-fr)*", "rm -rf"));
        assert_debug_snapshot!(is_regex("^f", "rm -rf"));
    }
}
