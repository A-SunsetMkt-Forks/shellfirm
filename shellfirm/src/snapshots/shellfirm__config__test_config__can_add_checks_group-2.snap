---
source: src/config.rs
expression: "settings_config.add_checks_group(&[\"base\".into()])"
---
Ok(
    Config {
        challenge: Math,
        includes: [
            "test",
            "base",
        ],
        version: "0.1.6",
        checks: [
            Check {
                test: "is value",
                method: Contains,
                enable: true,
                description: "description",
                from: "",
                challenge: Default,
                filters: {},
            },
            Check {
                test: ":(){ :|: & };:",
                method: Contains,
                enable: true,
                description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack.",
                from: "base",
                challenge: Default,
                filters: {},
            },
            Check {
                test: "\\s*history.*.[|].*.(bash|sh)",
                method: Regex,
                enable: true,
                description: "You are going to executes every command from the command log that you have already executed.",
                from: "base",
                challenge: Default,
                filters: {},
            },
            Check {
                test: "(\\.*)reboot[^.\"']*{0,}",
                method: Regex,
                enable: true,
                description: "You are going to reboot your machine.",
                from: "base",
                challenge: Default,
                filters: {},
            },
            Check {
                test: "(\\.*)shutdown[^.\"']*{0,}",
                method: Regex,
                enable: true,
                description: "You are going to shutdown your machine.",
                from: "base",
                challenge: Default,
                filters: {},
            },
        ],
    },
)
